### PR TITLE
Microsoft.Data.Sqlite: Support BigInteger, Int128 & UInt128

### LINKED ISSUE #30930: Microsoft.Data.Sqlite: Support BigInteger, Int128 & UInt128
Similar to how we handle `decimal`, we could read and write these values as TEXT.

### PR DESCRIPTION
**Pull request by smnsht**
- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [x] The code builds and tests pass locally (also verified by our automated build checks)
- [x] Commit messages follow this format:
```
        Summary of the changes
        - Detail 1
        - Detail 2
```
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code follows the same patterns and style as existing code in this repo

Fixes #30930

### CODE REVIEW

File: src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs [hunk 1]
```diff
@@ -140,6 +141,16 @@ public virtual long GetInt64(int ordinal)
                 : GetInt64Core(ordinal);
 
         protected abstract long GetInt64Core(int ordinal);
+#if NET7_0_OR_GREATER
+        public virtual Int128 GetInt128(int ordinal)
+            => (Int128)GetBigInteger(ordinal);
+        public virtual UInt128 GetUInt128(int ordinal)
+            => (UInt128)GetBigInteger(ordinal);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs:148]: I wouldn't go through BigInteger for these. Just read the string here too and parse it directly to these types.

File: src/EFCore.Sqlite.Core/Scaffolding/Internal/SqliteDatabaseModelFactory.cs [hunk 2]
```diff
@@ -745,6 +747,18 @@ protected virtual void InferClrTypes(DbConnection connection, DatabaseTable tabl
                     continue;
                 }
 
+                if (BigInteger.TryParse(max, NumberStyles.Float, null, out var maxBigInt)
+                        && BigInteger.TryParse(min, NumberStyles.Float, null, out var minBigInt))
```
Comment (bricelam) [inline src/EFCore.Sqlite.Core/Scaffolding/Internal/SqliteDatabaseModelFactory.cs:751]: This probably needs to move into the if block for decimals above.

Comment (smnsht) [reply src/EFCore.Sqlite.Core/Scaffolding/Internal/SqliteDatabaseModelFactory.cs:751]: **min** and **max** strings for big integers looks like this: 3.40282366920938e+38 or 3.40282366920939e+39 or -1.70141183460469e+38, etc...

Regex for decimal won't match this, so I came up with this solution.

Tested in method **Column_ClrType_is_set_when_big_integers** (SqliteDatabaseModelFactoryTest)

File: src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs (hunk 1 above)
Comment (smnsht) [reply src/Microsoft.Data.Sqlite.Core/SqliteValueReader.cs]: Fixed

**Comment by smnsht** [comment]
Added support for Int128/UInt128/BigInteger via converters

**Review by FirdavsAsadov — APPROVED** [review]

**Review by FirdavsAsadov — APPROVED** [review]

**Review by roji — COMMENTED** [review]
I'm generally skeptical of introducing this kind of support at the ADO.NET level (the EF level is a different story)... Yes, SQLite does provide storage in text fields for some types which aren't natively supported (e.g. DateTime); but that's done since those types are very fundamental, and without them the driver would be hard to use productively.

The types here (e.g. Int128, UInt128) don't fall into that category - no other provider supports them (because they aren't natively supported by databases), so I'm not sure why Microsoft.Data.Sqlite should. If go down this path, this begs the question of whether we should add native support e.g. for .NET Uri, IPAddress and other types in the ADO.NET provider, just to convert them to strings in the database.

Bottom line, it really seems reasonable to ask users to convert their Int128s into strings before asking M.D.Sqlite to save them etc.

(note that for BigInteger specifically, Npgsql does map that to a PG `decimal`, which is arbitrary-precision, and therefore faithfully represents the value (even if the other way doesn't fully work, since BigInteger can't represent fractions).

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs [hunk 3]
```diff
@@ -492,6 +493,44 @@ public override long GetInt64(int ordinal)
                     ? throw new InvalidOperationException(Resources.NoData)
                     : _record.GetInt64(ordinal);
 
+#if NET7_0_OR_GREATER
+        /// <summary>
+        ///     Gets the value of the specified column as a <see cref="Int128" />.
+        /// </summary>
+        /// <param name="ordinal">The zero-based column ordinal.</param>
+        /// <returns>The value of the column.</returns>
+        public Int128 GetInt128(int ordinal)
```
[the remark this answers is not in the document]
Comment (roji) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs:502]: One additional note: this textual representation of these number types means that values don't preserve ordering. In other words, if you save BigIntegers for 10 and 2, you get '10' and '2' in the database as strings, but '2' is bigger than '10' when doing lexicographic comparison.

At the very least, this means that we should block comparisons/ordering on the EF side (like we do for date/time textual representations); but I think it also makes it even more dubious to provide this as a default feature at the ADO.NET level.

**Comment by roji** [comment]
One additional note: this textual representation of these number types means that values don't preserve ordering. In other words, if you save BigIntegers for 10 and 2, you get '10' and '2' in the database as strings, but '2' is bigger than '10' when doing lexicographic comparison.

At the very least, this means that we should block comparisons/ordering on the EF side (like we do for date/time textual representations); but I think it also makes it even more dubious to provide this as a default feature at the ADO.NET level.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smnsht) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs:502]: As a "silent partner", I will just wait for consensus among maintainers.
I don't have anything else to add to this discussion.

**Comment by smnsht** [comment]
Anything from this pull request could be retained?

How about `StringNumberConverter` + its unit tests?
My changes in this converter are not tied to SQLite. If issue #30930 could be altered to something like  "Support BigInteger, Int128 & UInt128 at converter level", .....

**Comment by roji** [comment]
@smnsht we need to have a conversation about this as a team and decide which direction we want to take - please hang and we'll let you know.

**Comment by ajcvickers** [comment]
We discussed this again, and have decided not to take any of this, primarily for the reasons discussed above. I suspect we would take a PR to EF to add value converters for these, although there isn't complete agreement on the team on that.

Thanks again for your efforts here.
